### PR TITLE
manifest: Add submodule representation in as_dict

### LINKED ISSUE #739: west manifest freeze/resolve will lose submodules information
the project in raw west.yml
```yaml
  - name: test_repo
    path: manifest_repo/test_repo
    revision: main
    submodules:
    - path: modules/test_module
```
And run `west manifest --resolve -o freeze.ym`
freezed west.yml
```yaml
  - name: test_repo
    path: manifest_repo/test_repo
    revision: main
```
The submodules key gone.
Not sure whether this field was ignored in project's as_dict api: https://github.com/zephyrproject-rtos/west/blob/49f0a101df45ea735437bd388606a1219d9e76f5/src/west/manifest.py#L866

### PR DESCRIPTION
**Pull request by pdgendt**
Submodules were missing when calling --freeze or --resolve

Fixes #739

### CODE REVIEW

File: tests/test_project.py [hunk 1]
```diff
@@ -806,6 +807,28 @@ def test_update_submodules_list(repos_tmpdir):
...
+                    f'^    url: {re.escape(str(net_tools))}$',
+                    '^    revision: [a-f0-9]{40}$',
+                    '^    submodules:$',
+                    f'^    - path: {re.escape(str(kconfiglib_submodule))}$',
+                    '^      name: Kconfiglib$',
+                    '^  self:$',
+                    '^    path: mp$']
+
+    for eline_re, aline in zip(expected_res, actual):
+        assert re.match(eline_re, aline) is not None, (aline, eline_re)
```
Comment (marc-hb) [inline tests/test_project.py:830]: At first, this (wrongly) felt overkill. But now I think this is actually really nice and pretty generic. Can you "promote" this duplicated `for` loop to a named function? Something like `match_multiline_regex()` or something.

The only potential "indirection" issue I foresee: make sure it's still possible to quickly tell which test failed when one does. If needed then maybe let the `match_multiline_regex()` pass a (sub)test name as a string. Hopefully not even needed.

Oh, and you did check that `pytest` does not have this already, right?

File: src/west/manifest.py [hunk 2]
```diff
@@ -882,6 +882,15 @@ def as_dict(self) -> Dict:
                 _west_commands_maybe_delist(self.west_commands)
         if self.groups:
             ret['groups'] = self.groups
+        if isinstance(self.submodules, bool) and self.submodules:
+            ret['submodules'] = self.submodules
```
Comment (marc-hb) [inline src/west/manifest.py:886]:
```suggestion
            ret['submodules'] = True
```

**Review by marc-hb — COMMENTED** [review]
Minor suggestions below, everything else looks good, thanks!

File: src/west/manifest.py (hunk 2 above)
Comment (pdgendt) [reply src/west/manifest.py]: Updated

File: tests/test_project.py (hunk 1 above)
Comment (pdgendt) [reply tests/test_project.py]: I created a helper method for this, didn't find any `pytest` out-of-the-box helper.

**Review by marc-hb — APPROVED** [review]

**Pull request merged**